### ISSUE TITLE
Clean up after DROP SCHEMA on worker nodes

### ISSUE THREAD

**Issue by marcocitus** [citusdata/citus#5615]
We currently do not propagate DROP SCHEMA to worker nodes, but only drop the distributed tables. The reason is that propagating DROP SCHEMA creates self-deadlock scenarios when the transaction later tries to e.g. recreate the same schema. Since we do schema creation outside of the regular transaction (to support parallel create_distributed_table, propagating schemas with different owners), recreation would deadlock against the pending DROP SCHEMA.

As a consequence, we leak schemas on the worker nodes, especially for applications that frequently create and drop schemas (e.g. for ETL jobs). We should figure out a way to clean those up.

- [ ] Maybe propagate DROP SCHEMA if not in a transaction block
- [ ] Implement some method for dropping dropped schemas on worker nodes

**Comment by thanodnl** [comment]
We might need to reconsider our tradeoffs on parallelism here.

We might need to more start focussing on correctness vs parallelism. It is hard to make correct things fast, however it is way harder if not impossible to make fast-incorrect things correct.

Consequence would be that when users drop schema's, they either need to switch to sequential mode or error out if already using parallel connections. Wonder if we could/should make it easier to understand for users when sequential mode is switched to, and what implications it has for performance as well.

If we were able to switch to sequential mode when users drop schema's it will be easier to keep everything in sync.

**Comment by onderkalaci** [comment]
> We might need to reconsider our tradeoffs on parallelism here.

As far as I can see, that doesn't have anything to do with parallelism here. The main problem is that we send `CREATE SCHEMA` or any object creation over a separate super-user connection (see https://github.com/citusdata/citus/issues/5139). 

So, to make `DROP SCHEMA` transactional, we should send it over the `CurrentUser` initiated connection/remote transaction blocks.

If you do:
```SQL
BEGIN;
CREATE SCHEMA sc1;

-- create schema goes over superuser connection and commits -> we are good
CREATE TABLE sc1.test(a int);
SELECT create_distributed_table('sc1.test', 'a');

-- assume we send DROP SCHEMA, it should go over the current user connection for TX visibility & deadlock purposes
DROP SCHEMA sc1 CASCADE;

-- now, we are still in the transaction block
-- if the user tries to create the SAME schema AND it goes over 
CREATE SCHEMA sc1;
CREATE TABLE sc1.test(a int);

-- CREATE SCHEMA is send over a superuser connection
-- and it deadlocks with the DROP SCHEMA which has not been committed yet 
SELECT create_distributed_table('sc1.test', 'a');
```

> recreation would deadlock against the pending DROP SCHEMA.

Maybe we could try to prevent re-creating the same schema on the tx block. Re-creating the same schema doesn't seem very common thing? 

That should allow sending the DROP SCHEMA command over the same connection? 

The implementation could perhaps be naive, keep a simple linkedList where we keep all the names of the schemas created. If you attempt to create the same again, we give a nice error. If TX is done, we clear the list

**Comment by marcocitus** [comment]
> Maybe we could try to prevent re-creating the same schema on the tx block. Re-creating the same schema doesn't seem very common thing?

I think it's at least common to do:
```
BEGIN;
-- clean up from previous run
DROP SCHEMA IF EXISTS foo CASCADE;
CREATE SCHEMA foo;
...
```

**Comment by agedemenli** [comment]
We have discussed with Onder and thought about an abstract design plan, until we have decided that it's not a perfect option:

* We should propagate schemas as objects, and we should always propagate it, like types; unless it's created in a transaction block.
* We should detect the deadlock cases for transaction blocks, like in Marco's example, and provide a nice error message for those cases.
* We have tried to detect these cases in `create_distributed_table`. Other than it's a bit difficult to detect those cases, the main problem that we discovered is; other than explicitly executed `CREATE SCHEMA` commands and `create_distributed_table`, there are countless cases that we need to handle, because of the dependencies. Many types of objects could depend on schemas, such as tables, functions, statistcs', views etc. Therefore if we are going to continue with a solution like this, we might think on a more generic solution, like not in `create_distributed_table`, but in somewhere around `EnsureDependenciesExistOnAllNodes`


One other note: We might want to check the privileges granted to roles, instead of ownership. 
To summarize, on authorization, these are the factors that inflate the total number of combinations that we need to cover: [regular user, superuser], [has privilege, is owner, none], on [schema itself, an object that depends on the schema]. 
Though I suspect there's any difference between ownership/having privilege (Should we treat them in the same way ??)
I also wonder if there should be any differences between objects that depend on a schema when handling them in transaction block

**Comment by onderkalaci** [comment]
One pragmatic approach I start to consider is: On `EnsureDependenciesExistOnAllNodes` go over all the objects, and if all objects belong to the current user, create objects via the current user. It prevents majority of the edge cases.

The problems with that:  We still have the potential self-deadlocks if you re-create the same schema by another user in the same TX block. I think that sounds super unlikely. But, also we'd be breaking the users who already does that. 

 Although we have https://github.com/citusdata/citus/pull/3869 which prevent further edge cases if the command somehow reaches to the remote node via another user in the same TX, we still be vulnarable to self deadlocks with:
 ```SQL
 BEGIN;
SET ROLE another_role;
CREATE SCHEMA sc1;
SET ROLE current_user;

....

DROP SCHEMA current_user;

-- re-create the same schema with another user
SET ROLE another_role;
CREATE SCHEMA sc1;
SET ROLE current_user;

...
```

**Comment by onderkalaci** [comment]
What happens if we get into deadlocks? Is there a workaround?

**Comment by thanodnl** [comment]
> > We might need to reconsider our tradeoffs on parallelism here.
> 
> As far as I can see, that doesn't have anything to do with parallelism here. The main problem is that we send `CREATE SCHEMA` or any object creation over a separate super-user connection (see #5139).

That has exactly to do with parallelism.

To allow future parallelism we do not propagate node scoped object creation when in a transaction block. However, for most objects when you alter them, to maintain correctness we either switch to sequential mode or error out in the transaction block, with a hint to change to sequential mode before executing any statement.

> So, to make `DROP SCHEMA` transactional, we should send it over the CurrentUser initiated connection/remote transaction blocks.

I think the easiest to achieve this is, when dropping distributed schema's to take the same logic types take. Enforce sequential mode or error if already parallel and propagate the drop.

For creates we need to add an extra check to propagate directly over same user connection if the transaction is already in sequential mode. Inspiration from types:
 - https://github.com/citusdata/citus/blob/master/src/backend/distributed/commands/type.c#L1192-L1195
 - https://github.com/citusdata/citus/blob/master/src/backend/distributed/commands/type.c#L157-L161
 - https://github.com/citusdata/citus/blob/master/src/backend/distributed/commands/type.c#L408
 
 I don't think there is any other way to support deletes and creates of the same schema in transaction blocks, except for running in sequential mode.

**Comment by onderkalaci** [comment]
> That has exactly to do with parallelism.

We should definitely switch to sequential mode. But the problem is broader. See for example the below on the master branch where we start with local execution and get into self-deadlock due to superuser connections:
```SQL
BEGIN;
	SET LOCAL citus.multi_shard_modify_mode TO 'sequential';

	CREATE TYPE t1 AS (a int);
	CREATE TABLE test_1 (a int, b t1);
	SELECT create_distributed_table('test_1', 'a');

	DROP TABLE test_1;
	DROP TYPE t1;

	CREATE TYPE t1 AS (a int);
	CREATE TABLE test_2 (a int, b t1);
	SELECT create_distributed_table('test_2', 'a');
	ERROR:  canceling the transaction since it was involved in a distributed deadlock

```

This is the exact same problem we discuss here for `types`, however the point is that `CREATE SCHEMA` or `DROP SCHEMA` with the same name in a tx block are more likely to be used than other objects.

**Comment by agedemenli** [comment]
1. Add a check into `EnsureDependenciesExistOnAllNodes`, that checks if all the dependent objects belong to the current user or not. If so, use `SendMetadataCommandListToWorkerInCoordinatedTransaction` with the current user. If not, fallback to the current logic. For the *'if not'* case, we can consider the approach mentioned [here](https://github.com/citusdata/citus/issues/5139#issuecomment-1022039801), even though it seems a bit tricky and puzzling.

(The first item will, hopefully, fix #5655. 
And will fix most of the cases for this issue. Our main concern is not to break any user with these changes. On top of that, we can follow the next steps for this issue.)

2. Add schema propagation as a regular object.
3. Introduce a new GUC like `citus.unsafe_skip_drop_schema`, and check it when propagating `DROP SCHEMA` commands. Skip propagating if it's set to true. Will be false by default. With that, if we break any user, there will be an option to fallback to the old mechanism.
4. Unmark object distributed if an object is dropped on the worker, via `DROP SCHEMA CASCADE`.

**Comment by thanodnl** [comment]
> This is the exact same problem we discuss here for `types`, however the point is that `CREATE SCHEMA` or `DROP SCHEMA` with the same name in a tx block are more likely to be used than other objects such as types.

I ran the first part of your example transaction and I think the bug is _not_ related to using super using connections. But instead that the type is not propagated directly when running in sequential mode. I expected it to directly propagate in the transaction.

When directly propagating it uses the same connection as the DML statements, and thus, should not have deadlock potential.

_not_ connecting with super user when propagating missing dependencies via `EnsureDependenciesExistOnAllWorkers` will only solve issues if the dependencies are all created by the same user, which is only a subset of dependencies this function ultimately resolves. Instead it should propagate the dependencies as the current user via the same connections as the DML will use in sequential mode.

**Comment by thanodnl** [comment]
> ell, there are still problems about connecting with different users in a tx block (e.g., we can have multiple connections to the same node). I'm really not fan of allowing that (unless what we have today: commit immediately ).

I would separate this problem from the current problem at hand. Changing roles within a tx fails in many different ways at the moment, including isolation and consistency if not deadlocking, which deadlocking is arguable better compared to giving wrong answers.

This needs to be solved in a more holistic way by changing how we manage and cache connections. A backend keeps track of 4 different user id's I believe. This state should be recreated on the session we use for executing any statement. Not only DDL.


> The summary of Ahmet's proposal is that if the current user is sufficient to propagate all the dependencies, just do it. This allows us to handle majority of the cases for `CREATE/DROP SCHEMA` nicely.

I disagree with this. The only way we should be doing this is forcing sequential mode or error out if that is not possible anymore, which we already do with drop's for most objects, except for schema's. It is a bug that any create is not propagated when the transaction is already in sequential mode. When solving that bug we actually handle creates and drops correctly.

Changing the behaviour of `EnsureDependenciesExistsOnAllWorkers` in this way we will encounter more complex problems in the future due to the low exercise rate of creating non-current-user-owned objects. And not attending the root cause of our current problems.

**Comment by agedemenli** [comment]
> The only way we should be doing this is forcing sequential mode or error out if that is not possible anymore,

Wouldn't that break some of the users, e.g the ones that have such cases in their application tests?

I'm not against the idea of forcing sequential mode or erroring out; but in that case, shouldn't we suggest a workaround for the 'broken' users?

**Comment by thanodnl** [comment]
Not all usecases require sequential mode.

Simply creating a type and using it in a distributed table can work in parrallel execution.
However as soon as you want to delete a distributed object either before or after you should require sequential mode.

By propagating ddl when a transaction is already in sequential mode we allow users to perform more complex series of ddl statements in a transaction block which would now not be possible due to distributed deadlocks.

**Comment by onderkalaci** [comment]
I think all discussion comes down to: Do we want to break existing user with [these](https://github.com/citusdata/citus/issues/5615#issuecomment-1019909520) patterns or not? 

- If we are OK with breaking (e.g., requiring SET `LOCAL sequential_execution`) for such TX blocks, Nils' suggestion works fine. 
- If not, still first follow Nils' suggestion. On top of that, we should eliminate the need to the extra connection by another user , and call `EnsureSequentialModeForTypeDDL()` (or such) right before that current user connection is used. With that, we eliminate the user explicitly setting the execution mode.

I do see both as useful things by their own. But, it might be fine to skip the second one and take this small risk of breaking users with such patterns.

**Comment by onderkalaci** [comment]
I think all discussion comes down to: Do we want to break existing user with [these](https://github.com/citusdata/citus/issues/5615#issuecomment-1019909520) patterns or not? 

- If we are OK with breaking (e.g., requiring explicit SET `LOCAL sequential_execution` by the user) for such TX blocks, Nils' suggestion works fine. 
- If not, still first follow Nils' suggestion. On top of that, we should eliminate the need to the extra connection by another user  as we would not be automatically propagating on `CREATE SCHEMA` , and call `EnsureSequentialModeForTypeDDL()` (or such) right before that current user connection is used. With that, we eliminate the user explicitly setting the execution mode.

I do see both as useful things by their own. But, it might be fine to skip the second one and take this small risk of breaking users with such patterns.

**Comment by onderkalaci** [comment]
Ah, so given that `CREATE/DROP SCHEMA` doesn't require parallelism anyway, we can directly call `EnsureSequentialModeForTypeDDL ` ?

**Comment by thanodnl** [comment]
> Ah, so given that `CREATE/DROP SCHEMA` doesn't require parallelism anyway, we can directly call `EnsureSequentialModeForTypeDDL ` ? 

The parallelism is not directly related to the DDL, more to any DML in the same transaction.

> Even before `ShouldPropagateXXX()` function? We normally do that afterwards

I think we should first check the should propagate predicate. It is (partially) controlled by GUC's that advanced users can use to disable certain behaviour (eg, propagating creates - although many usecases will not work ofc when you disable) by Ensuring Sequential before these policy checks we disallow people to stay in parallel mode by disabling propagation of certain objects. Hence we ensure sequential only after we have decided we need to propagate the ddl statement.
